s3torchbenchmarking/src/s3torchbenchmarking/dcp/benchmark.py

```diff
@@ -0,0 +1,156 @@
+        # diff(max, min)) metric.
+        collector: List[float] = []
+        while not save_timestamps.empty():
+            collector.append(save_timestamps.get())
+        ptp_save_times.append(np.ptp(collector))
+
+    save_results(
+        cfg,
+        benchmark_model,
+        ptp_save_times=ptp_save_times,
+        processing_times=processing_times,
+        monitor=monitor,
+    )
+
+
+def validate_world_size(world_size: int) -> int:
+    """Enforce `world_size` to be within the current node's capacity.
+
+    FIXME: only works when the backend is "nccl"; what about "gloo" and CPUs?
+    """
+    device_count = torch.cuda.device_count()
+    if world_size > device_count:
+        logger.warning(
+            "Received a `world_size` of %i, while only %i are available: decreasing to %i",
+            world_size,
+            device_count,
+            device_count,
+        )
+        return device_count
+    return world_size
+
+
+def get_writer(cfg: DictConfig, suffix: str) -> FileSystemWriter:
+    """Instantiate a checkpoint writer based on the input config."""
+    if cfg.checkpoint.storage == "disk":
+        local_path = Path(cfg.path) / suffix
+        logger.info("Saving checkpoint to %s (local disk)...", local_path)
+        return dcp.FileSystemWriter(local_path, thread_count=cfg.thread_count)
+    elif cfg.checkpoint.storage == "s3":
+        uri = build_checkpoint_uri(cfg.s3.uri, suffix)
+        logger.info("Saving checkpoint to %s (S3)...", uri)
+        return S3StorageWriter(cfg.s3.region, uri, thread_count=cfg.thread_count)
+    raise ValueError(f"Storage writer {cfg.checkpoint.storage} not supported")
+
+
+def build_checkpoint_uri(s3_uri: str, suffix: str) -> str:
+    suffix = suffix.lstrip("/")
+    return s3_uri + suffix if s3_uri.endswith("/") else s3_uri + "/" + suffix
+
+
+def setup(backend: str, world_size: int, rank: int) -> None:
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+    dist.init_process_group(backend=backend, world_size=world_size, rank=rank)
+
+
+def run(
+    rank: int,  # needs to be passed first (provided by `multiprocessing.spawn` automatically)
+    backend: str,
+    epochs: int,
+    model: Module,
+    world_size: int,
+    storage_writer: FileSystemWriter,
+    save_timestamps: Queue,
+) -> None:
+    """Execute the actual code for checkpoint saving.
+
+    This function is meant to be executed in subprocesses."""
+
+    # FIXME: configure logging in subprocess accordingly
+
+    torch.cuda.set_device(rank)
+    setup(backend=backend, world_size=world_size, rank=rank)
+    rank = dist.get_rank()
+
+    device_id = rank % torch.cuda.device_count()
+    model.to(device_id)
+
+    ddp_model = DistributedDataParallel(model, device_ids=[device_id])
+
+    start_time = perf_counter()
+    for i in range(epochs):
+        dcp.save(ddp_model.state_dict(), storage_writer=storage_writer)
+    end_time = perf_counter()
```

Comment:
Let's think about edge cases, the start and end time here is still impacted by the load time of the model and other processes going on before this, we should remove the time taken for all the processes happening before this step, one approach I tried is

def run(
    rank: int,
    model: Module,
    world_size: int,
    storage_writer: FileSystemWriter,
    shared_start_times: list,
    shared_end_times: list,
):
    start_time = perf_counter()
    logger.info("Running on rank %s...", rank)

    #torch.cuda.set_device(rank)
    setup(rank, world_size)
    rank = dist.get_rank()
    device_id = rank % torch.cuda.device_count()
    torch.cuda.set_device(device_id)

    logger.debug("Device ID: %s", device_id)

    model.to(device=torch.device("cpu"))
    wrapped_model = DistributedDataParallel(model)

    logger.info("Saving checkpoint on rank %s...", rank)

    model_load_end = perf_counter()

    dcp.save(
        wrapped_model.state_dict(),
        storage_writer=storage_writer,
    )
    end_time = perf_counter() - (model_load_end - start_time)